### ISSUE TITLE
Refactoring the code to complete C++ version

### ISSUE THREAD

**Issue by YeZhao** [jgeisler0303/DDP-Generator#1]
Hi @jgeisler0303, I just realized this DDP implementation and it looks nice work. I wonder whether there is any demo code that is fully written in C++. I am trying to refactor this code by removing the Maxima part. I guess a few functions are generated and complied by the make_iLQG.m file. Do you have a sense of the flexibility for creating such a full C++ version? Thank you, -- Ye

**Comment by jgeisler0303** [comment]
Hi,
I attached the generated files for the CarParking example so you may see from these the API/interface for the problem specific code: [Car_gen_files.zip](https://github.com/jgeisler0303/DDP-Generator/files/786112/Car_gen_files.zip)

Indeed I did start to refactor the code to C++. But more importantly I have to separate branches, one with OpenMP support and one which uses Eigen3 for the matrix operations. I always wanted to merge these two and then go on from there with the refactoring. But I somehow got stuck with other work. If you are interested, I could push my latest work to a dev branch. It's still built around the idea of using maxima generated problem specific code but the interface changed a bit (the data structure is a bit more like a class and a bit less like a struct :-)). But that should be no big concern for you if you want to decouple the core code from the maxima part anyway (which I think is a good idea). Last time I touched the project I was fairly certain that all changes work without problems.

**Comment by YeZhao** [comment]
Thank you for this timely response, @jgeisler0303. The reason why I am considering to decouple the iLQG code from the maxima is that I want to devise MPC algorithms based on DDP. This traj opt algorithm is expected to be integraed into a manipulator. I wonder how is the current status of the dev branch code. 

Another issue is when I run make_iLQG.m, the following error pops up in maxima_errors.log, 

```/bin/bash: /home/yezhao/Downloads/maxima-5.39.0: Is a directory```

Only err.log and maxima_errors.log are generated in the Car_gen_files folder (no iLQG_func.c and ILQG_problem.h are created yet). Do you have any idea what potentially causes this? I think I install maxima successfully as you instructed (replace the files in gentran folder by yours, and follow the maxima instructions).

**Comment by YeZhao** [comment]
Fix the previous path setting issue! Now I got another error. When the programs run 

```batchload("genenerator_main.mac");``` 

in make_iLQG.mac file, the following error shows up

```apply: found false where a function was expected.```

Then I dig into the details of genenerator_main.mac and realized that the following line (Line 548) causes the issue
 ```hu_: makelist(subvar(hu, i_work-1, i_), i_, 0, nu-1)```

where hu is a Lisp array generated from make_array.

I also tried 

```(%i1) x : foo $

(%i2) i : 3 $

(%i3) subvar (x, i);
(%o3)                         foo
                                 3
(%i4) foo : [aa, bb, cc, dd, ee]$

(%i5) subvar (x, i);
(%o5)                          cc
(%i6) arraymake (x, [i]);
(%o6)                         foo
                                 3
(%i7) ''%;
(%o7)                          cc
```

this case works. And this indicates that subvar should work fine with an ordinary array I guess. But it throws the error when I try subvar(hu, 1). So I believe that subvar on a Lisp array might be related to the error.

Now I am not quite sure how to fix this issue. Do you have any good suggestions? Thank you,

**Comment by jgeisler0303** [comment]
Which problem definition were you trying to make? It seem s hu is not defined. What do `member(h, arrays)` and `arrayinfo(hu);` say?

**Comment by jgeisler0303** [comment]
I have now pushed my latest developments to the `openmp_eigen3` branch. I think, last time I worked on it I was fairly happy with the results comparing the old code against openmp code and eigen3.

The CarParking results differed a bit when using the Eigen3 code but that can be explained by numerical differences.

With the OpenMP code, I was not sure yet, if it gave any speed improvements. But the performance gain will probably only show for systems with higher state dimensions.

 The different specializations are arranged into so-called targets. Each target only adds those files that are different from the `common` target which is basically the old code you already know. Further, each target has its own make-script that "knows" how it has to be built. All the special make-scripts have an identical interface are to be called by the common `make_target` script telling it the problem to make, the concrete target and additional target specific options.

So, for example if you wanted to make the CarParking problem with the eigen3 target you have to start octave, `addpath('PATH_TO_DDDP_GENERATOR');`, cd into the CarParking example directory, and then `make_target('optDefCar', 'ddp/eigen', '-DDEBUG_BACKPASS=1 -DDEBUG_FORWARDPASS=1 -DFULL_DDP=0', 0)`.

**Comment by YeZhao** [comment]
I try to set up the car parking example. This is the output

```
(%i621) arrayinfo(hu);
(%o621)                      [declared, 2, [3, 1]]
(%i622) member(h,arrays);
(%o622)                              true
```

**Comment by YeZhao** [comment]
I wonder whether I miss to install certain packages and set up dependencies. You mentioned 
``
On my Linux computer I got gentran only to work using Steel Bank Common Lisp (SBCL) implementation. 
``
Does this require some extra installments? When I configure maxima, I get the following message such that I assume SBCL is already all set-up. Does this make sense?

```
Summary:
SBCL enabled. Executable name: "sbcl"
default lisp: sbcl
wish executable name: "wish"
```

**Comment by jgeisler0303** [comment]
Your output indicates that everything should be fine and that you are running Steel Bank Common Lisp; I guess on Windows? 

Are you sure the problem comes from that line? The error message implies that the problem stems from the first argument to the `apply` function is not as expected (like this: `apply(false, []);`). There are several calls to `apply`. One is in line 213: do you have a variable `properties` set (it shouldn't be)? Other candidates are `defrule`, `apply1`, `gradef`.

I will try this tomorrow at work as I cannot replicate the problem here an my Linux machine.

**Comment by YeZhao** [comment]
Thank you. As a quick reply, I use Ubuntu 14.04 and Matlab R2014a. I am trying to check whether I can set up the OpenMP code.

**Comment by YeZhao** [comment]
I run the OpenMP code in Octave (before I use Matlab). It seems the same error appears. Here I attach the output

```
octave:17> make_target('optDefCar', 'ddp/eigen', '-DDEBUG_BACKPASS=1 -DDEBUG_FORWARDPASS=1 -DFULL_DDP=0', 0)
Maxima said:
(%i1) problem_file:"/home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/optDefCar.mac"
(%o1) /home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/o\
ptDefCar.mac
(%i2) target_dir:"/home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen"
(%o2) /home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/C\
ar_ddp_eigen
(%i3) batchload("make_iLQG.mac")
using states:  [x_, y_, t, v] 
using inputs:  [w, a] 
apply: found false where a function was expected.
 -- an error. To debug this try: debugmode(true);


Compiling iterate.cpp
g++ -c -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/octave-3.8.1/octave/.. -I/usr/include/octave-3.8.1/octave -I/usr/include/mpi  -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -pthread -fopenmp   -I. -I/home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/  -DDEBUG_BACKPASS=1 -DDEBUG_FORWARDPASS=1 -DFULL_DDP=0 iterate.cpp -o /home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/obj/iterate.o

Compiling iLQG_mex.cpp
g++ -c -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/octave-3.8.1/octave/.. -I/usr/include/octave-3.8.1/octave -I/usr/include/mpi  -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -pthread -fopenmp   -I. -I/home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/  -DDEBUG_BACKPASS=1 -DDEBUG_FORWARDPASS=1 -DFULL_DDP=0 iLQG_mex.cpp -o /home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/obj/iLQG_mex.o

Compiling line_search.cpp
g++ -c -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/octave-3.8.1/octave/.. -I/usr/include/octave-3.8.1/octave -I/usr/include/mpi  -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -pthread -fopenmp   -I. -I/home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/  -DDEBUG_BACKPASS=1 -DDEBUG_FORWARDPASS=1 -DFULL_DDP=0 line_search.cpp -o /home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/obj/line_search.o

Compiling back_pass.cpp
g++ -c -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/octave-3.8.1/octave/.. -I/usr/include/octave-3.8.1/octave -I/usr/include/mpi  -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -pthread -fopenmp   -I. -I/home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/  -DDEBUG_BACKPASS=1 -DDEBUG_FORWARDPASS=1 -DFULL_DDP=0 back_pass.cpp -o /home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/obj/back_pass.o

Compiling boxQP.cpp
g++ -c -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/octave-3.8.1/octave/.. -I/usr/include/octave-3.8.1/octave -I/usr/include/mpi  -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -pthread -fopenmp   -I. -I/home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/  -DDEBUG_BACKPASS=1 -DDEBUG_FORWARDPASS=1 -DFULL_DDP=0 boxQP.cpp -o /home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/obj/boxQP.o

Compiling printMat.cpp
g++ -c -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/octave-3.8.1/octave/.. -I/usr/include/octave-3.8.1/octave -I/usr/include/mpi  -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -pthread -fopenmp   -I. -I/home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/  -DDEBUG_BACKPASS=1 -DDEBUG_FORWARDPASS=1 -DFULL_DDP=0 printMat.cpp -o /home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/obj/printMat.o

Compiling iLQG.cpp
g++ -c -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/octave-3.8.1/octave/.. -I/usr/include/octave-3.8.1/octave -I/usr/include/mpi  -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -pthread -fopenmp   -I. -I/home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/  -DDEBUG_BACKPASS=1 -DDEBUG_FORWARDPASS=1 -DFULL_DDP=0 iLQG.cpp -o /home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/obj/iLQG.o

Compiling /home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/iLQG_func.cpp
g++ -c -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/octave-3.8.1/octave/.. -I/usr/include/octave-3.8.1/octave -I/usr/include/mpi  -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -pthread -fopenmp   -I. -I/home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/  -DDEBUG_BACKPASS=1 -DDEBUG_FORWARDPASS=1 -DFULL_DDP=0 /home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/iLQG_func.cpp -o /home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/obj/iLQG_func.o

There were compile errors logged to file: /home/yezhao/Downloads/DDP-Generator-openmp_eigen3/examples/CarParking/Car_ddp_eigen/compile_err.log
```

**Comment by jgeisler0303** [comment]
That is very strange. Still the same error with apply. I have the exact same Ubuntu. What does your maxima say? Mine says:
`
Maxima 5.37.2 http://maxima.sourceforge.net                                                                         
using Lisp SBCL 1.1.14.debian
`

Could you run the contents of "make_iLQG.mac" by pasting it line by line into the maxima command line (set problem_file, target_dir appropriately).

**Comment by YeZhao** [comment]
In Matlab, maxima says basically the same error message:

```
Maxima said:
(%i1) problem_file:"/home/yezhao/DDP-Generator-master/optDefCar.mac"
(%o1)           /home/yezhao/DDP-Generator-master/optDefCar.mac
(%i2) target_dir:"/home/yezhao/DDP-Generator-master/Car_gen_files"
(%o2)           /home/yezhao/DDP-Generator-master/Car_gen_files
(%i3) batchload("make_iLQG.mac")
using states:  [x_, y_, t, v] 
using inputs:  [w, a] 
apply: found false where a function was expected.
 -- an error. To debug this try: debugmode(true);
```

And I print out the ``command_str`` and run it in maxima command line (set problem_file and target_dir are set as shown above). The same error still exists.

**Comment by YeZhao** [comment]
The maxima version I use is 5.39.0
```
yezhao@falconer ~ $ maxima -v
+ [ sbcl = clisp ]
+ [ sbcl = cmucl ]
+ [ sbcl = scl ]
+ [ sbcl = gcl ]
+ [ sbcl = acl ]
+ [ sbcl = openmcl ]
+ [ sbcl = ccl64 ]
+ [ sbcl = ecl ]
+ [ sbcl = sbcl ]
+ [ -x /usr/local/lib/maxima/5.39.0/binary-sbcl/maxima ]
+ exec sbcl --core /usr/local/lib/maxima/5.39.0/binary-sbcl/maxima.core --noinform --end-runtime-options --eval (cl-user::run) --end-toplevel-options -v
Maxima 5.39.0 http://maxima.sourceforge.net
using Lisp SBCL 1.1.14.debian
Distributed under the GNU Public License. See the file COPYING.
Dedicated to the memory of William Schelter.
The function bug_report() provides bug reporting information.
```

**Comment by jgeisler0303** [comment]
It was a problem with `subvar` that only showed in 5.39. I fixed it and pushed to github. Compared to code generated with 5.37 it seems identical although parameter sorting has changed.

**Comment by YeZhao** [comment]
Thanks a lot, @jgeisler0303. I'll give a try and let you know how it goes.

**Comment by jgeisler0303** [comment]
Hey @YeZhao , how's it going?

**Comment by YeZhao** [comment]
Hi @jgeisler0303, sorry about this huge delay. Yeah, the code works fine now. Thank you for the asking.

BTW, i am curious what propose the following parameters are used for (which seems not included in Yuval's matlab code) Thanks.
```
o->w_pen_init_l= 1.0;
    o->w_pen_init_f= 1.0;
    o->w_pen_max_l= INF;
    o->w_pen_max_f= INF;
    o->w_pen_fact1= 4.0; // 4...10 Bertsekas p. 123
o->w_pen_fact2= 1.0;
```

**Comment by jgeisler0303** [comment]
These parameters belong to a feature that bridges the one gap that Tassa's code leaves: state constraints. I implemented a method called [Augmented Lagrangian Method](https://en.wikipedia.org/wiki/Augmented_Lagrangian_method). I worked by [Constrained Optimization and Lagrange Multiplier Methods](http://www.mit.edu/~dimitrib/Constrained-Opt.pdf). It's not fully tested yet but the examples of the Brachistochrone wouldn't work without it.

There are methods for stage-wise  equality (hle) and inequality (hli) constraints and for terminal equality (hfe) and inequality (hfi) constraints. What still need work is the optimization of when to update the multipliers and when to update the penalty-weights (w_pen_...) that you found.

With code did you base your refactorization on, the master branch, the eigen code or the openmp one? How did you go about the class design? Did you use virtual methods for the problem-specific code or pointers to functions? Did you template the class with respect to the number of states and inputs? Where do you have your code? You are aware that this work is under the GPL and any modifications must be published under the same terms?

**Comment by YeZhao** [comment]
Thanks for the explanations! It looks an interesting problem to schedule penalty weights and Lagrange multipliers (maybe another parameter optimization in the outermost loop). 

Since I originally had the mex compile issues as mentioned above, I basically developed the DDP algorithm from scratch (mainly stemming from Yuval's matlab code). My code refers to a few code blocks from your algorithm. The algorithm is still under development. I also have a UDP version (a.k.a. unscented dynamic programming, which gets rid of computationally-heavy dynamics derivative). Here is a [pointer](https://github.com/YeZhao/drake/tree/mpc-ddp-dev/drake/examples/kuka_iiwa_arm/DDP_traj_gen) of my code, which I developed for our kuka arm.

Initially, I do have a base class named as `dynamicmodel`, and it declares a bunch of virtual functions, which are implemented by the derived class `kuka_arm`. But I realized it is not very helpful at this moment since we currently focus a lot on the manipulator system only. Later, if more robotic systems are applied, I can quickly switch it back. The class has not been templated yet. As the code becomes mature, I'll certainly consider the GPL matter and cite your code properly. Will be happy to discuss any terms and restrictions that should be followed.

Another question, in the README, it states the multi-threading seems not improving speed on your machine. Could you comment more on that part? Since my manipulator dynamics update occupies a large percentage of the total solver time (around 76%), I am interested in paralleling the computation for that part.

**Comment by jgeisler0303** [comment]
I think there are some rules of thumb, when to update. But I'm not sure how it should play with the regularization.

If you basically started from scratch, I guess there should be no licence issue. But thanks for sharing your code. I will certainly have a look.

Especially the UDP seems very interesting. I read a paper about it. But I had the feeling that finding the sigma-points via the inverse of the Hessian might also be quite computationally expensive. 

I see two points where multi-threading might be useful:
1.  The derivatives of stage k are not needed until the backpass reaches stage k. So, depending on the ratio of computational time for the derivatives compared to that of one backpass stage, one or more threads could compute the derivatives just ahead of the backpass loop. The backpass cannot be distributed over several threads because calculating each stage requires the result of the preceding one. In my code, I had the feeling that synchronising the threads and making sure the backpass doesn't overtake the derivatives computation is quite time consuming and thus only pays off when the time spent per stage is high enough.
1. Several line searches could be done in parallel. But that only saves time when more than one searches are needed. Otherwise it might even be harmful. Anyway, every search thread should be immediately terminated as soon as the first one finds a sufficient reduction of the cost.

Finally, if you are looking for super fast yet easy to implement MPC, you should also take a look at the [ACADO Toolkit](http://acado.github.io/). You can define your problem definition in MATLAB and get highly optimized custom code including a choice of special integrators and solvers.

**Comment by YeZhao** [comment]
Great to know these details! I'll check more details how the multi-threading is implemented. So if points 1&2 are implemented properly, the computation time should be reduced, right? Is your current code trying to target this?

Essentially, UDP generates very similar trajectories as DDP/iLQR. Its computation time per iteration approaches more to that of iLQR while its convergence rate is more similar to that of DDP. For my kuka arm problem, the most computationally heavy part comes from the state update where a RK4 method is used. In each function evaluation for next stage state, the rigid body tree computation takes a lot of time (it is basically forward dynamics. The kinematics update and matrix inverse are time-consuming). So the computational problem is not directly related which the optimization algorithm is chosen. It occurs when the algorithm is applied to problems with scaled-up dimensionality.

On first sight, ACADO looks a nice tool. I will learn more about it.

**Comment by jgeisler0303** [comment]
Points 1&2 are implemented in the files in "target/ddp/mp" folder. It is based on OpenMP. It's a while ago though. So, I don't remember if there were any problems.

**Comment by YeZhao** [comment]
Thanks! I will check it.

**Comment by YeZhao** [comment]
@jgeisler0303 I parallelized the sigma points calculation part by 20 threads (in my case, only standard C++ thread is used, and sigma point calculation costs a majority of the solve time). The total solver time now is two times faster than that of the single thread version. I realized that `pthread` is used in your master branch while `OpenMP` is used in the openmp_eigen3 branch. I searched online for some performance comparison of these two thread mechanisms. Especially for this DDP algorithm, I wonder whether you expect any speed performance improvement by using OpenMP.

**Comment by jgeisler0303** [comment]
I tried OpenMP because it seems to be more readable and structured. And it also enables quick changes to see how different loop parallelization strategies (fixed allocation, automatic allocation, blocked, etc.) affect the speed.
